src/array/from_geoarrow.rs

```diff
@@ -1,85 +1,94 @@
+    ToCellListArray, ToCellsOptions,
 };
 use crate::algorithm::CompactOp;
 use crate::array::from_geo::geometry_to_cells;
-use crate::array::{CellIndexArray, H3ListArray, H3ListArrayBuilder};
+use crate::array::{CellIndexArray, H3ListArray};
 use crate::error::Error;
+use arrow::array::OffsetSizeTrait;
 use geo_types::Geometry;
 use geoarrow::array::WKBArray;
+use geoarrow::trait_::GeometryArrayAccessor;
 use geoarrow::GeometryArrayTrait;
-use geozero::ToGeo;
 use h3o::CellIndex;
 #[cfg(feature = "rayon")]
 use rayon::prelude::{IntoParallelIterator, ParallelIterator};
 
-macro_rules! impl_from_geoarrow {
-    ($($array_type:ty),*) => {
-        $(
-        impl ToCellListArray for $array_type {
-            fn to_celllistarray(&self, options: &ToCellsOptions) -> Result<H3ListArray<CellIndex>, Error> {
+macro_rules! impl_to_cells {
+    ($array_type:ty, $offset:tt) => {
+        impl<$offset: OffsetSizeTrait> ToCellListArray<$offset> for $array_type {
+            fn to_celllistarray(
+                &self,
+                options: &ToCellsOptions,
+            ) -> Result<H3ListArray<CellIndex, $offset>, Error> {
                 self.iter_geo()
                     .map(|v| v.map(Geometry::from))
                     .to_celllistarray(options)
             }
         }
 
-        impl ToCellIndexArray for $array_type {
+        impl<$offset: OffsetSizeTrait> ToCellIndexArray for $array_type {
             fn to_cellindexarray(&self, options: &ToCellsOptions) -> Result<CellIndexArray, Error> {
                 self.iter_geo()
                     .map(|v| v.map(Geometry::from))
                     .to_cellindexarray(options)
             }
         }
+    };
+    ($array_type:ty) => {
+        impl<O: OffsetSizeTrait> ToCellListArray<O> for $array_type {
+            fn to_celllistarray(
+                &self,
+                options: &ToCellsOptions,
+            ) -> Result<H3ListArray<CellIndex, O>, Error> {
+                self.iter_geo()
+                    .map(|v| v.map(Geometry::from))
+                    .to_celllistarray(options)
+            }
+        }
 
-        )*
+        impl ToCellIndexArray for $array_type {
+            fn to_cellindexarray(&self, options: &ToCellsOptions) -> Result<CellIndexArray, Error> {
+                self.iter_geo()
+                    .map(|v| v.map(Geometry::from))
+                    .to_cellindexarray(options)
+            }
+        }
     };
 }
 
-impl_from_geoarrow!(
-    geoarrow::array::LineStringArray,
-    geoarrow::array::MultiLineStringArray,
-    geoarrow::array::MultiPointArray,
-    geoarrow::array::MultiPolygonArray,
-    geoarrow::array::PointArray,
-    geoarrow::array::PolygonArray
-);
+impl_to_cells!(geoarrow::array::LineStringArray<O>, O);
+impl_to_cells!(geoarrow::array::MultiLineStringArray<O>, O);
+impl_to_cells!(geoarrow::array::MultiPointArray<O>, O);
+impl_to_cells!(geoarrow::array::MultiPolygonArray<O>, O);
+impl_to_cells!(geoarrow::array::PointArray);
```

Comment:
Interesting! I've been wondering how to implement this! So far I've had separate hard-coded impls for PointArray https://github.com/geoarrow/geoarrow-rs/blob/3a2aaa883126274037cabaf46b1f5f6459938297/src/algorithm/geo/area.rs#L52-L63

I suppose this is slightly better, though you still have to duplicate the implementation in the macro.

Reply:
There still may be better ways to implement this. I just try to keep macro implementations not too complicated so I still understand them after coming back to them after a while ;)